### PR TITLE
Drop Rails 4.2 support

### LINKED ISSUE #468: Remove support for Rails 4.2
In general: Remove rails 4.2 support from gem

## Summary

Remove rails 4.2 from gem support. Requires 3 things.
- rails dropping support for 4.2 Or a period of around 3-6months from below PR
- Extension to ruby support (Currently being worked on in #466)
- Code cleaning around conditionals that support 4.2 (I'm not sure on how extensive this is, but I know it exists).

Once this work has been done, we would then look to "soon" after remove 5.0 support.

## Expected Behavior

5.0, 5.1, 5.2, 6.0 all supported

## Current Behavior

4.2, 5.0, 5.1, 5.2, 6.0 all supported

## Possible Solution

Fix stuff

## Steps to Reproduce (for bugs)

Each PR to be done separate, culmination in a release of 2.1

## Context & Motivation

Tidy up of old codebase

## Your Environment

N/A

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
## Summary

Drop Rails 4.2 support. This PR closes #468.

## Details

Grep for 4.2 and change/remove the stuff.

## Motivation and Context

It's no longer supported by the Rails team, and it hasn't been for a while.

## How Has This Been Tested?

Not yet, I'll watch CI directly.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist:

- [ ] I've added tests for my code.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by olleolleolle** [comment]
The failing job, I restarted it - this was the error:

```
0.07s$ bin/install_geckodriver.sh

+curl --silent --show-error --location --fail --retry 3 --output /tmp/geckodriver_linux64.tar.gz https://github.com/mozilla/geckodriver/releases/download/v0.24.0/geckodriver-v0.24.0-linux64.tar.gz

curl: (22) The requested URL returned error: 429 too many requests
```

File: Rakefile [hunk 1]
```diff
@@ -26,11 +26,11 @@ namespace :test do
     Rake::Task['appraisal'].invoke('test')
   end
 
-  desc 'Run tests against specified gemfile, e.g. rake test:gemfile[rails_3_0]'
+  desc 'Run tests against specified gemfile, e.g. rake test:gemfile[rails_6_0]'
```
Comment (olleolleolle) [inline Rakefile:29]: This was a good find!

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -29,8 +29,8 @@ This document is a guide for those maintaining Cucumber-Rails, and others who wo
     bundle install
     bin/install_geckodriver.sh
     bin/install_webpacker.sh
-    # Then to run tests on one specific gemfile (i.e. Rails 4.2), run
-    bundle exec appraisal rails_4_2 rake test
+    # Then to run tests on one specific gemfile (i.e. Rails 6.0), run
```
Comment (olleolleolle) [inline CONTRIBUTING.md:32]:
```suggestion
    # Then to run tests on one version-specific Gemfile (e.g. gemfiles/rails_6_0.gemfile), run
```

Wording like this also shows concretely where the individual file is located. Perhaps good?

Comment (deivid-rodriguez) [reply CONTRIBUTING.md:32]: Much much better :+1:

**Comment by luke-hill** [comment]
Ideally I'd want to wait a bit after doing @mvz PR first, so lets just get some feedback on this one and also try push the other one along.

**Comment by deivid-rodriguez** [comment]
@luke-hill Isn't it better the other way around? This inmediately reduces the size of our CI matrix and somewhat resolves https://github.com/cucumber/cucumber-rails/pull/466#discussion_r436250935.

**Comment by luke-hill** [comment]
Not quite, as the ruby update is something as a whole the repo / community has been moving towards in the last 6months-1year. Removing rails is something a bit chunkier. I'd rather remove rails after we've done the ruby update and ideally as a separate cut, so we've got more flexibility for users.

That being said, there's no reason it has to be a long time in between the releases.

**Comment by deivid-rodriguez** [comment]
Oh, so you're not talking about the order the PRs are merged in, but about leaving dropping Rails support for a later release rather than for the next? Then I disagree even more :stuck_out_tongue_closed_eyes:. In my opinion, there's no significant differences between dropping support for Rails versions that reached their end of life and dropping support for ruby versions that reached their end of life.

In any case, no big deal. Merge this PR at your own pace :+1:.

**Comment by deivid-rodriguez** [comment]
@luke-hill I rebased this for you, so it's ready to merge whenever you're ready.

**Comment by luke-hill** [comment]
Going forward (From v2.1, which I've just cut, but not yet published), I'm gonna try convince @olleolleolle and @mvz (If they're cool), to adopt the cucumber / cucumber-ruby approach of having people write their own changelogs to save us hassle.

Once We've merged the release branch (Which fixes all the old stuff), can you rebase this again and then write a changelog entry for this>? Happy for it to be merged once that's all done, as it then gets a nice style going forwards.

**Comment by deivid-rodriguez** [comment]
Sure, just ping me and I'll rebase and add a changelog entry.

**Comment by mvz** [comment]
> I'm gonna try convince @olleolleolle and @mvz (If they're happy), to adopt the cucumber / cucumber-ruby approach of having people write their own changelogs to save us hassle.

That makes a lot of sense to me, so consider me convinced. :slightly_smiling_face:

**Comment by olleolleolle** [comment]
Yeah please go ahead with this!

**Comment by luke-hill** [comment]
@deivid-rodriguez Can you rebase now and add a changelog entry (Changelog now is a lot more fully formed), add it in the unreleased part at the top for a 2.2 (or 3.0), cut

**Comment by deivid-rodriguez** [comment]
Sure, let me do that now.

By the way, is something holding off the 2.1 release? It's a bit confusing that a tag has been cut and a release published on Github, but the release is not yet available on rubygems.org.

**Comment by deivid-rodriguez** [comment]
Done!

**Pull request merged**